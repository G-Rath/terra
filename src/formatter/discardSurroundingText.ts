import { TFFileContents } from '@src/types';

export type DiscardSurroundingTextBehaviour =
  | 'except-comments'
  | 'except-comments-and-newlines'
  | 'all';

/**
 * Discards the surrounding text of the given `contents`, based on the specified behaviour:
 *
 *  * `all`: all surrounding text will be discarded
 *  * `except-comments`: only surrounding text containing comments will be kept
 *  * `except-comments-and-newlines`: only surrounding text containing either at
 *                                    least one newline, or a comment will be kept
 *
 * @param {TFFileContents} contents
 * @param {DiscardSurroundingTextBehaviour} discardBehaviour
 *
 * @return {TFFileContents}
 */
export const discardSurroundingText = (
  contents: TFFileContents,
  discardBehaviour: DiscardSurroundingTextBehaviour
): TFFileContents =>
  JSON.parse(
    JSON.stringify(contents, (key, value) => {
      if (key !== 'surroundingText') {
        return value;
      }

      const surroundingText: Record<string, string> = { ...value };

      Object.keys(value).forEach(k => {
        surroundingText[k] = '';

        if (discardBehaviour !== 'all') {
          const text: string = value[k];

          if (/#|\*\/|\/\*|\/\//u.test(text)) {
            surroundingText[k] = text;
          }

          if (discardBehaviour === 'except-comments-and-newlines') {
            if (text.includes('\n')) {
              surroundingText[k] = text;
            }
          }
        }
      });

      return surroundingText;
    })
  );
