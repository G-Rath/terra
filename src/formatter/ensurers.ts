import {
  TokenType,
  ensureTextStartsWithTokens,
  parseSurroundingText,
  printTokens
} from '@src/formatter';
import { TFBlock } from '@src/types';

export type Ensurer = (blocks: TFBlock[]) => TFBlock[];

export const ensureTopLevelBlocksAreSeparated: Ensurer = blocks => {
  if (blocks.length <= 1) {
    return blocks;
  }

  return [
    blocks[0],
    ...blocks.slice(1).map(block => {
      const { leadingOuterText } = block.surroundingText;

      const parsedText = parseSurroundingText(leadingOuterText);

      const firstNewlineTokenIndex = parsedText.findIndex(
        token => token.type === TokenType.Newline
      );

      if (firstNewlineTokenIndex === -1) {
        parsedText.splice(
          firstNewlineTokenIndex,
          0,
          { type: TokenType.Newline },
          { type: TokenType.Newline }
        );
      } else if (
        parsedText.length <= firstNewlineTokenIndex + 1 ||
        parsedText[firstNewlineTokenIndex + 1].type !== TokenType.Newline
      ) {
        parsedText.splice(firstNewlineTokenIndex, 0, {
          type: TokenType.Newline
        });
      }

      return {
        ...block,
        surroundingText: {
          ...block.surroundingText,
          leadingOuterText: printTokens(parsedText)
        }
      };
    })
  ];
};

export const ensureLabelsHaveLeadingSpace: Ensurer = blocks =>
  blocks.map(block => ({
    ...block,
    labels: block.labels.map(label => ({
      ...label,
      surroundingText: {
        ...label.surroundingText,
        leadingOuterText: ensureTextStartsWithTokens(
          label.surroundingText.leadingOuterText,
          [{ type: TokenType.Whitespace, content: ' ' }]
        )
      }
    }))
  }));

export const ensurers: readonly Ensurer[] = Object.freeze([
  ensureTopLevelBlocksAreSeparated,
  ensureLabelsHaveLeadingSpace
]);
