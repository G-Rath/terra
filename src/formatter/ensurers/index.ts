export * from './ensureTopLevelBlocksAreSeparated';
export * from './ensureNoSpacesAfterFunctionName';
export * from './ensureSpaceBeforeOpeningBrace';
export * from './ensureLabelsHaveLeadingSpace';
export * from './ensureClosingBraceOnNewline';
export * from './ensureOneArgumentPerLine';
export * from './ensureIndentation';
