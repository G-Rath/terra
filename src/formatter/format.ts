import { TFFileContents } from '@src/types';

const cloneBlocks = (
  blocks: TFFileContents['blocks']
): TFFileContents['blocks'] => JSON.parse(JSON.stringify(blocks));

export const format = (fileContents: TFFileContents): TFFileContents => {
  if (fileContents.blocks.length === 0) {
    return {
      blocks: cloneBlocks(fileContents.blocks),
      surroundingText: { ...fileContents.surroundingText }
    };
  }

  return {
    blocks: cloneBlocks(fileContents.blocks),
    surroundingText: { ...fileContents.surroundingText }
  };
};
