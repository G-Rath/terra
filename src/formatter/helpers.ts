import {
  Token,
  ensureClosingBraceOnNewline,
  ensureIndentation,
  ensureLabelsHaveLeadingSpace,
  ensureNoSpacesAfterFunctionName,
  ensureSpaceBeforeOpeningBrace,
  ensureTopLevelBlocksAreSeparated,
  parseSurroundingText,
  printTokens
} from '@src/formatter';
import {
  SurroundingInnerText,
  SurroundingOuterText,
  TFBlock
} from '@src/types';

export interface NodeWithOuterText {
  surroundingText: SurroundingOuterText;
}

export interface NodeWithInnerText {
  surroundingText: SurroundingInnerText;
}

export type Ensurer = (blocks: TFBlock[]) => TFBlock[];

export const ensurers: readonly Ensurer[] = Object.freeze([
  ensureTopLevelBlocksAreSeparated,
  ensureNoSpacesAfterFunctionName,
  ensureSpaceBeforeOpeningBrace,
  ensureLabelsHaveLeadingSpace,
  ensureClosingBraceOnNewline,
  ensureIndentation
]);

/**
 * Compares the given `Token`s to see if they are equal.
 *
 * A `Token` is equal to another if both `Token`s have the same type & content.
 *
 * @param {Token & {content?: any}} a
 * @param {Token & {content?: any}} b
 *
 * @returns {boolean}
 */
const areTokensEqual = (
  a: Token & { content?: unknown },
  b: Token & { content?: unknown }
): boolean => a.type === b.type && a.content === b.content;

/**
 * Ensures that the given array of `tokens` ends with the second given array of `Token`s.
 *
 * This compares in the same manner as `String#endsWith`.
 *
 * @param {readonly Token[]} tokens
 * @param {readonly Token[]} tokensToEndWith
 *
 * @returns {readonly Token[]}
 */
const ensureTokensEndWithTokens = (
  tokens: readonly Token[],
  tokensToEndWith: readonly Token[]
): readonly Token[] => {
  if (tokens.length === 0) {
    return tokensToEndWith;
  }

  if (
    tokensToEndWith.length <= tokens.length &&
    tokensToEndWith.every((token, i) =>
      areTokensEqual(token, tokens[tokens.length - i - 1])
    )
  ) {
    return tokens;
  }

  return tokens.concat(tokensToEndWith);
};

export const ensureTextEndsWithTokens = (
  text: string,
  tokens: readonly Token[]
): string =>
  printTokens(ensureTokensEndWithTokens(parseSurroundingText(text), tokens));

/**
 * Ensures that the given array of `tokens` starts with the second given array of `Token`s.
 *
 * This compares in the same manner as `String#startsWith`.
 *
 * @param {readonly Token[]} tokens
 * @param {readonly Token[]} tokensToStartWith
 *
 * @returns {readonly Token[]}
 */
const ensureTokensStartWithTokens = (
  tokens: readonly Token[],
  tokensToStartWith: readonly Token[]
): readonly Token[] => {
  if (tokens.length === 0) {
    return tokensToStartWith;
  }

  if (
    tokensToStartWith.length <= tokens.length &&
    tokensToStartWith.every((token, i) => areTokensEqual(token, tokens[i]))
  ) {
    return tokens;
  }

  return tokensToStartWith.concat(tokens);
};

export const ensureTextStartsWithTokens = (
  text: string,
  tokens: readonly Token[]
): string =>
  printTokens(ensureTokensStartWithTokens(parseSurroundingText(text), tokens));
