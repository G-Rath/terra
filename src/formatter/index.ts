export * from './parseSurroundingText';
export * from './printTokens';
export * from './format';
