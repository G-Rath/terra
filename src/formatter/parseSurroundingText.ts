import { StringCursor } from '@src/parser';

export enum TokenType {
  Whitespace = 'whitespace',
  Newline = 'newline',
  Comment = 'comment'
}

export interface WhitespaceToken {
  type: TokenType.Whitespace;
  content: string;
}

export interface NewlineToken {
  type: TokenType.Newline;
}

export interface CommentToken {
  type: TokenType.Comment;
  content: string;
}

export type Token = WhitespaceToken | NewlineToken | CommentToken;

const parseComment = (cursor: StringCursor): string => {
  const multiLineComment = cursor.char === '/' && cursor.peek() === '*';
  let comment = '';

  do {
    comment += cursor.advance();
  } while (
    !(cursor.isAtEnd() || comment.endsWith(multiLineComment ? '*/' : '\n'))
  );

  if (!multiLineComment && comment.endsWith('\n')) {
    comment = comment.slice(0, -1);
    cursor.rewind(1);
  }

  return comment;
};

const buildToken = (cursor: StringCursor): Token => {
  const char = cursor.advance();

  switch (char) {
    case '\n':
      return { type: TokenType.Newline };
    case ' ':
      return {
        type: TokenType.Whitespace,
        content: char
      };
    case '#':
    case '/':
      cursor.rewind(1);

      return {
        type: TokenType.Comment,
        content: parseComment(cursor)
      };
    default:
      throw new Error(`unknown token "${char}"`);
  }
};

export const parseSurroundingText = (text: string): Token[] => {
  const cursor = new StringCursor(text);
  const tokens: Token[] = [];

  while (!cursor.isAtEnd()) {
    tokens.push(buildToken(cursor));
  }

  return tokens;
};
