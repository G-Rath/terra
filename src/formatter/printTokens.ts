import { Token, TokenType } from '@src/formatter';

export const printTokens = (tokens: readonly Token[]): string =>
  tokens
    .map(token => {
      switch (token.type) {
        case TokenType.Newline:
          return '\n';
        case TokenType.Comment:
        case TokenType.Whitespace:
          return token.content;
        default:
          throw new Error('unknown token type');
      }
    })
    .join('');
