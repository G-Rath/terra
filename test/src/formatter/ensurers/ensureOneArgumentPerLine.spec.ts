import { ensureOneArgumentPerLine } from '@src/formatter';
import { makeFormatter } from '@test/helpers';
import { irlFixtures } from '@test/irlFixtures';

describe('ensureOneArgumentPerLine', () => {
  describe('irl fixtures', () => {
    const removeNewlines = (str: string): string =>
      str
        .split('\n')
        .map(line => line.trim())
        .join(' ');

    const ensureOneArgumentPerLineFormatter = (str: string): string =>
      makeFormatter(ensureOneArgumentPerLine)(removeNewlines(str));

    it.each(irlFixtures)('it formats this %s as expected', (_, content) => {
      try {
        // eslint-disable-next-line jest/no-restricted-matchers
        expect(ensureOneArgumentPerLineFormatter(content)).toMatchSnapshot();
      } catch (error) {
        console.log(content);

        throw error;
      }
    });
  });
});
