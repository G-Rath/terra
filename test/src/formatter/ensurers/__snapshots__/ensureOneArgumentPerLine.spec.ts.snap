// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ensureOneArgumentPerLine irl fixtures it formats this awesome locals block as expected 1`] = `
"
locals {
 admin_locations = [ {
 cidr_range  = \\"xx.xxx.xxx.xxx/xx\\",
 description = \\"My Office In A City\\" } ] }"
`;

exports[`ensureOneArgumentPerLine irl fixtures it formats this cool aws_dlm_lifecycle_policy resource as expected 1`] = `
"
resource \\"aws_dlm_lifecycle_policy\\" \\"server-data\\" {
 description        = \\"Server Data Volume lifecycle policy\\"
 execution_role_arn = aws_iam_role.dlm.arn
 state              = \\"ENABLED\\" 
 policy_details {
 resource_types = [\\"VOLUME\\"] 
 schedule {
 name = \\"\\\\\${var.days_to_keep_snapshots} days of server snapshots\\" 
 create_rule {
 interval      = 24
 interval_unit = \\"HOURS\\" 
 times = [\\"14:00\\"] } 
 retain_rule {
 count = var.days_to_keep_snapshots } 
 tags_to_add = {
 SnapshotCreator = \\"DLM\\" } 
 copy_tags = true } 
 target_tags = {
 \\"DLM-SnapshotGroup\\" = \\"server-data\\" } } }"
`;

exports[`ensureOneArgumentPerLine irl fixtures it formats this jaw dropping aws_db_subnet_group resource as expected 1`] = `
"
resource \\"aws_db_subnet_group\\" \\"my_db\\" {
 name        = \\"\\\\\${local.env_prefix}-db-subnets\\"
 subnet_ids  = module.production_network.public_subnet_ids
 description = \\"Subnets which may contain DB instances within \\\\\${local.env_prefix} VPC\\" 
 tags = merge( local.common_tags, {
 \\"Name\\" = \\"\\\\\${local.env_prefix}-db-subnets\\" }, ) }"
`;

exports[`ensureOneArgumentPerLine irl fixtures it formats this radical aws_iam_policy_document resource as expected 1`] = `
"
data \\"aws_iam_policy_document\\" \\"dlm\\" {
 statement {
 effect  = \\"Allow\\"
 actions = [ \\"ec2:CreateSnapshot\\", \\"ec2:DeleteSnapshot\\", \\"ec2:DescribeVolumes\\", \\"ec2:DescribeSnapshots\\" ] 
 resources = [\\"*\\"] } 
 statement {
 effect  = \\"Allow\\"
 actions = [ \\"ec2:CreateTags\\" ] 
 resources = [\\"arn:aws:ec2:*::snapshot/*\\"] } }"
`;

exports[`ensureOneArgumentPerLine irl fixtures it formats this simple provider block as expected 1`] = `
"
provider \\"aws\\" {
 version    = \\"~> 2.30\\"
 access_key = var.AWS_ACCESS_KEY_ID
 secret_key = var.AWS_SECRET_ACCESS_KEY
 region     = \\"ap-southeast-2\\" }"
`;

exports[`ensureOneArgumentPerLine irl fixtures it formats this single line data block as expected 1`] = `
"
data \\"aws_caller_identity\\" \\"current\\" {}"
`;
